src/diffusers/models/embeddings.py

```diff
@@ -683,3 +718,110 @@ def forward(
+        )
+        self.frequency_embedding_size = frequency_embedding_size
+        self.outdim = hidden_size
+
+    def forward(self, size: torch.Tensor, batch_size: int):
+        if size.ndim == 1:
+            size = size[:, None]
+
+        if size.shape[0] != batch_size:
+            size = size.repeat(batch_size // size.shape[0], 1)
+            assert size.shape[0] == batch_size
+        current_batch_size, dims = size.shape[0], size.shape[1]
+        size = size.reshape(-1)
+        size_freq = get_timestep_embedding(
+            size, self.frequency_embedding_size, downscale_freq_shift=0, flip_sin_to_cos=True
+        ).to(size.dtype)
+
+        size_emb = self.mlp(size_freq)
+        size_emb = size_emb.reshape(current_batch_size, dims * self.outdim)
+        return size_emb
+
+
+class CombinedTimestepSizeEmbeddings(nn.Module):
+    """
+    For PixArt-Alpha.
+
+    Reference:
+    https://github.com/PixArt-alpha/PixArt-alpha/blob/0f55e922376d8b797edd44d25d0e7464b260dcab/diffusion/model/nets/PixArtMS.py#L164C9-L168C29
+    """
+
+    def __init__(self, embedding_dim, size_emb_dim, use_additional_conditions: bool = False):
+        super().__init__()
+
+        self.time_proj = Timesteps(num_channels=256, flip_sin_to_cos=True, downscale_freq_shift=0)
+        self.timestep_embedder = TimestepEmbedding(in_channels=256, time_embed_dim=embedding_dim)
+
+        self.use_additional_conditions = use_additional_conditions
+        if use_additional_conditions:
+            self.use_additional_conditions = True
+            self.resolution_embedder = SizeEmbedder(size_emb_dim)
+            self.aspect_ratio_embedder = SizeEmbedder(size_emb_dim)
+
+    def forward(self, timestep, resolution, aspect_ratio, batch_size, hidden_dtype):
+        timesteps_proj = self.time_proj(timestep)
+        timesteps_emb = self.timestep_embedder(timesteps_proj.to(dtype=hidden_dtype))  # (N, D)
+
+        if self.use_additional_conditions:
+            resolution = self.resolution_embedder(resolution, batch_size=batch_size)
+            aspect_ratio = self.aspect_ratio_embedder(aspect_ratio, batch_size=batch_size)
+            conditioning = timesteps_emb + torch.cat([resolution, aspect_ratio], dim=1)
+        else:
+            conditioning = timesteps_emb
+
+        return conditioning
+
+
+class CaptionProjection(nn.Module):
+    """
+    Projects caption embeddings. Also handles dropout for classifier-free guidance.
+
+    Adapted from https://github.com/PixArt-alpha/PixArt-alpha/blob/master/diffusion/model/nets/PixArt_blocks.py
+    """
+
+    def __init__(self, in_features, hidden_size, class_dropout_prob, num_tokens=120):
+        super().__init__()
+        self.mlp = nn.Sequential(
+            nn.Linear(in_features=in_features, out_features=hidden_size, bias=True),
+            nn.GELU(approximate="tanh"),
+            nn.Linear(in_features=hidden_size, out_features=hidden_size, bias=True),
+        )
```

Comment:
```suggestion
        self.linear_1 = ...
        self.linear_2 = ....
```

Let's rewrite this since we'll probably want to be able to re-use a "caption-proj" module